### ISSUE TITLE
transformer: babel-plugin-transform-modules-commonjs

### ISSUE THREAD

**Issue by Boshen** [oxc-project/oxc#4050]
https://babeljs.io/docs/babel-plugin-transform-modules-commonjs

For Rolldown app mode, we need to custom runtime to execute module.

**Comment by DonIsaac** [comment]
This would be great to have! I need it for oxc-jest.

**Comment by 7086cmd** [comment]
I may attempt to implement it when I have some free time (although I am currently quite occupied). Given my familiarity with the format logic in Rolldown, perhaps we could simply adapt the `render_chunk_export` function in Rolldown here?

In my opinion, this functionality should be prioritized after ES2015 (? not sure).

**Comment by Boshen** [comment]
> In my opinion, this functionality should be prioritized after ES2015 (? not sure).

Implementing this would unblock a lot of things, for example `oxc-jest`.

> perhaps we could simply adapt the render_chunk_export function in Rolldown here?

You may want to copy it over to complete the implementation first. We can look at deduplicating code later.

**Comment by 7086cmd** [comment]
> Implementing this would unblock a lot of things, for example oxc-jest.

I mean the order of transforming, the current order is `x0_`, `x1_`, etc. I think it should be after the currently last (ES2015)?

> You may want to copy it over to complete the implementation first. We can look at deduplicating code later.

OK, but the Rolldown uses its own logic in handling modules and focuses on dependencies, yet this task's essential thing is the syntax transforming (from import / export decl to call expr), IMO.

**Comment by Boshen** [comment]
> I mean the order of transforming, the current order is x0_, x1_, etc. I think it should be after the currently last (ES2015)?

Yes it should be the last visit. I wonder whether this should be a separate ast path though, what do you think @Dunqing ?

**Comment by Dunqing** [comment]
> Yes it should be the last visit.

In my opinion, should be visiting before lowering(esxxxx) plugins and after `TypeScript` and `React` plugins. Because this plugin may also contain high-level syntax after transformation

I wonder whether this should be a separate ast path though, what do you think @Dunqing ?

Yes, it should be

**Comment by overlookmotel** [comment]
The part that I can imagine being tricky is that other plugins may insert either `import`s or `require`s depending on the output format. e.g. React transform does this.

If those plugins know at outset that output format is CommonJS, some will be able to create tighter `require` code than what you get from inserting `import` statements initially and then transforming them to CommonJS after. For example React transform knows that `jsx` is not a mutable binding, so it can generate this:

```js
var _jsx = require("react/jsx-runtime").jsx;
_jsx(Foo, {});
```

instead of ending up with:

```js
var _jsxRuntime = require("react/jsx-runtime");
(0, _jsxRuntime.jsx)(Foo, {});
```

Same will be true for all imports of helper functions.

So it may be best to transform `import` to `require` before anything else, and then all other transforms generate `require`s if output format is CommonJS.

That would also cover Dunqing's point:

> In my opinion, should be visiting before... Because this plugin may also contain high-level syntax after transformation

**Comment by 7086cmd** [comment]
Yes, for now I put the plugin after react.

**Comment by christianscott** [comment]
Hi @Boshen. Would you consider reopening this issue? I am thinking about using `oxc` as our canonical ts-to-js transpiler and being able to transform modules to commonjs is a requirement for use with tools like jest (we don't have any plans to introduce support for running esm in jest).

**Comment by Boshen** [comment]
Reopening for migrating jest and other old usages over.
